llvm/lib/Target/AIE/AIEBaseAliasAnalysis.cpp

```diff
@@ -117,13 +152,281 @@ static const Value *getUnderlyingObjectAIE(const Value *V) {
+static bool
+isAddressingIntrinsicOutput(ArrayRef<const Value *> AdressingOutputs,
+                            const Value *Intrinsic) {
+
+  for (const Value *Counter : AdressingOutputs) {
+    if (Intrinsic != lookThroughAIEAddressingModes(Counter))
+      return false;
+  }
+
+  return true;
+}
+
+/// This function returns some tracking information related to the upward path
+/// that starts from a Value *TargetPointer to the Phi node that firstly defines
+/// a pointer value to the same ambiguous memory. We start the search from the
+/// last update of the pointer.
+std::optional<IntrinsicChainInfo>
+trackIntrinsicChain(const Value *LastUpdate, const Value *TargetPointer) {
+  IntrinsicChainInfo ChainInfo;
+  // We count backwards.
+  unsigned LastCalls = 0;
+  // Let's expose the real pointer.
+  TargetPointer = skipCast(TargetPointer);
+  for (unsigned Count = 0; Count < BaseObjectSearchLimit; ++Count) {
+
+    LastUpdate = skipCast(LastUpdate);
+    if (!LastUpdate->getType()->isPointerTy())
+      return std::nullopt; // not a pointer.
+
+    if (auto *GEP = dyn_cast<GEPOperator>(LastUpdate)) {
+      if (!GEP->hasAllZeroIndices())
+        return std::nullopt; // No pointer arithmetic is allowed.
+      // The result pointer and the first operand have
+      // the same value
+      LastUpdate = GEP->getPointerOperand();
+    } else if (auto *PHI = dyn_cast<PHINode>(LastUpdate)) {
+      // Reached final destination.
+      ChainInfo.IntrinsicCalls = ChainInfo.TotalIntrinsicCalls - LastCalls;
+      return ChainInfo;
+    } else if (auto *AIEObject = lookThroughAIEAddressingModes(LastUpdate)) {
+      IntrinsicArgsInfo NewArgSet = collectIntrinsicOperands(LastUpdate);
+      // Initialize counters.
+      if (!ChainInfo.IntrinsicParameters.CounterArgSet.empty() &&
+          !isAddressingIntrinsicOutput(
+              ArrayRef(ChainInfo.IntrinsicParameters.CounterArgSet),
+              AIEObject)) {
+        // if we already have counters, we need to reach the same
+        // AIEObject from the counters. Otherwise we can have some undesired
+        // counter update in between.
+        return std::nullopt; // not reachable.
+      }
+
+      if (!ChainInfo.IntrinsicParameters.ArgSet.empty() &&
+          ChainInfo.IntrinsicParameters.ArgSet != NewArgSet.ArgSet) {
+        // Constant iterator information mismatch between sequence of
+        // intrinsic calls. We stop here.
+        return std::nullopt;
+      }
+
+      // Update params for the next round.
+      ChainInfo.IntrinsicParameters.CounterArgSet = NewArgSet.CounterArgSet;
+      ChainInfo.IntrinsicParameters.ArgSet = NewArgSet.ArgSet;
+
+      LastUpdate = AIEObject;
+      ChainInfo.TotalIntrinsicCalls++;
+    } else {
+      return std::nullopt; // something that we can't understand.
+    }
+    // This information is crucial to calculate the number
+    // of pointer updates before TargetPointer.
+    if (LastUpdate == TargetPointer)
```

Comment:
Should we actually check `skipCast(LastUpdate) == TargetPointer`?

Reply:
Could you maybe add a test where there is a pointer cast right before the addressing intrinsic?

Reply:
Done! `@with_addrspace_cast`.